### PR TITLE
Add support for LetsEncrypt.

### PR DESCRIPTION
**Pull request by cellofellow**
This change is [Reviewable](https://reviewable.io/reviews/cesanta/docker_auth/105)

### CODE REVIEW

**Comment by rojer** [comment]
oh, nice, thanks for this!

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[auth_server/main.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cesanta/docker_auth/105#-KKuADLEOyUGgCRubw6j:-KKuADLEOyUGgCRubw6k:156056707) ([raw file](https://github.com/cesanta/docker_auth/blob/d6658ca0c823f871d525c47f8f56596a46a5488f/auth_server/main.go#L85)):*
> ```Go
> 		}
> 		if !m.Registered() {
> 			m.Register(c.Server.LetsEncrypt.Email, nil)
> ```

handle return value from Register here

---

*[auth_server/main.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cesanta/docker_auth/105#-KKuAI_Soc6Rqip5Zvuj:-KKuAI_Soc6Rqip5Zvuk:-740776332) ([raw file](https://github.com/cesanta/docker_auth/blob/d6658ca0c823f871d525c47f8f56596a46a5488f/auth_server/main.go#L87)):*
> ```Go
> 			m.Register(c.Server.LetsEncrypt.Email, nil)
> 		}
> 		tlsConfig = &tls.Config{
> ```

to reduce the amount of c&p, move setting TLS parameters into a helper function

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cesanta/docker_auth/105#-:-KKuAAghYDLSPme7jRkG:-1074837448)*

**Comment by raphink** [comment]
Cool beans. Any news on this?

**Comment by SydOps** [comment]
@cellofellow 

Could you give a sample on how to prepare below cachefile?

    cachefile: "/path/to/cachefile.letsencrypt"

**Comment by cellofellow** [comment]
How would you like the example? It just needs to be a path that the process can write to, creating a file holding keys and login information.

File: auth_server/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM busybox
+FROM golang
```
Comment (rojer) [inline auth_server/Dockerfile]: please leave this change out

File: examples/reference.yml [hunk 2]
```diff
@@ -15,6 +15,10 @@ server:  # Server settings.
   # TLS certificate and key.
   certificate: "/path/to/server.pem"
   key: "/path/to/server.key"
+  # Alternately, you may use LetsEncrypt
```
Comment (rojer) [inline examples/reference.yml]: to clarify - in this case, using separate token.{certificate,key} is a must, yes?

Comment (rojer) [inline examples/reference.yml]: also, for those not familiar with letsencrypt, give a link to the documentation.
the package doc (https://godoc.org/rsc.io/letsencrypt) gives a nice overview

**Comment by tlvenn** [comment]
@cellofellow any update on this ?

**Comment by techknowlogick** [comment]
For anyone coming here to use LE with this project, what I've done as a workaround is to use a self signed cert for the auth_server and then to put Caddy in front of auth server which handles LE nicely.

Example here: https://github.com/techknowlogick/docker-registry/blob/master/caddy/Caddyfile

**Comment by rojer** [comment]
so, i have some personal experience with LE now.
LE is awesome, i'm a big fan! i promise some movement on this no later than i need to renew our own cert ;)

**Comment by rojer** [comment]
this inspired https://github.com/cesanta/docker_auth/commit/a76426387685c721962e84e29f60c85456b28e86

thank you!

**Pull request closed without merging**